go/tools/builders/cgo2.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2019 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// cgo2.go provides new cgo functionality for use by the GoCompilePkg action.
+// We can't use the functionality in cgo.go, since it relies too heavily
+// on logic in cgo.bzl. Ideally, we'd be able to replace cgo.go with this
+// file eventually, but not until Bazel gives us enough toolchain information
+// to compile ObjC files.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// cgo2 processes a set of mixed source files with cgo.
+func cgo2(goenv *env, goSrcs, cgoSrcs, cSrcs, cxxSrcs, sSrcs, hSrcs []string, packagePath, packageName string, cc string, cppFlags, cFlags, cxxFlags, ldFlags []string, cgoExportHPath string) (srcDir string, allGoSrcs, cObjs []string, err error) {
+	// Report an error if the C/C++ toolchain wasn't configured.
+	if cc == "" {
+		err := cgoError(cgoSrcs[:])
+		err = append(err, cSrcs...)
+		err = append(err, cxxSrcs...)
+		err = append(err, sSrcs...)
+		return "", nil, nil, err
+	}
+
+	// If we only have C/C++ sources without cgo, just compile and pack them
+	// without generating code. The Go command forbids this, but we've
+	// historically allowed it.
+	// TODO(jayconrod): this doesn't write CGO_LDFLAGS into the archive. We
+	// might miss dependencies like -lstdc++ if they aren't referenced in
+	// some other way.
+	if len(cgoSrcs) == 0 {
```

Comment:
why not run them through `cc_library` ?

Reply:
I'm not sure I follow?

With this change, we won't need `cc_library` at all anymore. `go_library` will generate one action. It won't need to be a macro that generates a constellation of `filegroup`, `cc_library`, `cgo_codegen` and other rules anymore.

I think right thing to do is synthesize a .go file with `//cgo:linkflag` (or whatever the special pragma is). The Go compiler and linker will know what to do with those.

Reply:
sorry, I meant `cc_common.create_library_to_link`. It seems you are sending the cflags and then invoking `CC` manually ?

Reply:
I think we need to do it this way. This case is needed when there are .c files but no .go files with `import "C"`. There may have been .go files with `import "C"` that were filtered out, and there must be other pure .go files (we will synthesize one if none pass the filter). So the output is still a Go package, not a C library, and we need to read the files in order to decide how to build it.